### ISSUE TITLE
Error on union type in catch block

### ISSUE THREAD

**Issue by GoogleCodeExporter** [typetools/checker-framework#145]
```
I think this problem is caused by the new feature of union types in catch block 
parameters in Java 7 
(http://docs.oracle.com/javase/specs/jls/se7/html/jls-14.html#jls-14.20).


What steps will reproduce the problem?
1. Compile the attached program TestUnion.java using a command line similar to 
the attached shell script local_compile.sh

What is the expected output? What do you see instead?

I expected no warnings.  I saw this error and stack trace:

error: AnnotatedTypeMirror.createType: unidentified type 
java.lang.IndexOutOfBoundsException

  Exception: java.lang.Throwable: checkers.source.SourceChecker$CheckerError.<init>(SourceChecker.java:235)
  checkers.types.AnnotatedTypeMirror.createType(AnnotatedTypeMirror.java:90)
  checkers.types.AnnotatedTypeFactory.toAnnotatedType(AnnotatedTypeFactory.java:1270)
  checkers.types.AnnotatedTypeFactory.type(AnnotatedTypeFactory.java:1286)
  checkers.types.TypeFromTree$TypeFromTypeTree.visitUnionType(TypeFromTree.java:577)
  checkers.types.TypeFromTree$TypeFromTypeTree.visitUnionType(TypeFromTree.java:403)
  com.sun.tools.javac.tree.JCTree$JCTypeUnion.accept(JCTree.java:2069)
  com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:52)
  checkers.types.AnnotatedTypeFactory.fromTreeWithVisitor(AnnotatedTypeFactory.java:486)
  checkers.types.AnnotatedTypeFactory.fromTypeTree(AnnotatedTypeFactory.java:435)
  checkers.types.TypeFromTree$TypeFromMember.visitVariable(TypeFromTree.java:337)
  checkers.types.TypeFromTree$TypeFromMember.visitVariable(TypeFromTree.java:329)
  com.sun.tools.javac.tree.JCTree$JCVariableDecl.accept(JCTree.java:792)
  com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:52)
  checkers.types.AnnotatedTypeFactory.fromTreeWithVisitor(AnnotatedTypeFactory.java:486)
  checkers.types.AnnotatedTypeFactory.fromMember(AnnotatedTypeFactory.java:399)
  checkers.types.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:276)
  checkers.flow.DefaultFlow.newVar(DefaultFlow.java:55)
  checkers.flow.AbstractFlow.visitVariable(AbstractFlow.java:496)
  checkers.flow.AbstractFlow.visitVariable(AbstractFlow.java:65)
  com.sun.tools.javac.tree.JCTree$JCVariableDecl.accept(JCTree.java:792)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:67)
  checkers.flow.AbstractFlow.scan(AbstractFlow.java:187)
  checkers.flow.AbstractFlow.scan(AbstractFlow.java:65)
  com.sun.source.util.TreeScanner.visitCatch(TreeScanner.java:221)
  com.sun.tools.javac.tree.JCTree$JCCatch.accept(JCTree.java:1140)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:67)
  checkers.flow.AbstractFlow.scan(AbstractFlow.java:187)
  checkers.flow.AbstractFlow.visitTry(AbstractFlow.java:856)
  checkers.flow.AbstractFlow.visitTry(AbstractFlow.java:65)
  com.sun.tools.javac.tree.JCTree$JCTry.accept(JCTree.java:1110)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:67)
  checkers.flow.AbstractFlow.scan(AbstractFlow.java:187)
  checkers.flow.AbstractFlow.scan(AbstractFlow.java:65)
  com.sun.source.util.TreeScanner.scan(TreeScanner.java:90)
  com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:160)
  checkers.flow.AbstractFlow.visitBlock(AbstractFlow.java:910)
  checkers.flow.AbstractFlow.visitBlock(AbstractFlow.java:65)
  com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:846)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:67)
  checkers.flow.AbstractFlow.scan(AbstractFlow.java:187)
  checkers.flow.AbstractFlow.scan(AbstractFlow.java:65)
  com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:80)
  com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:143)
  checkers.flow.AbstractFlow.visitMethod(AbstractFlow.java:925)
  checkers.flow.AbstractFlow.visitMethod(AbstractFlow.java:65)
  com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:746)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:67)
  checkers.flow.AbstractFlow.scan(AbstractFlow.java:187)
  checkers.flow.AbstractFlow.visitClass(AbstractFlow.java:434)
  checkers.flow.AbstractFlow.visitClass(AbstractFlow.java:65)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:670)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:49)
  checkers.flow.AbstractFlow.scan(AbstractFlow.java:184)
  checkers.flow.AbstractFlow.scan(AbstractFlow.java:178)
  checkers.types.BasicAnnotatedTypeFactory.annotateImplicit(BasicAnnotatedTypeFactory.java:182)
  checkers.types.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:291)
  checkers.types.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:1149)
  checkers.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:140)
  checkers.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:76)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:670)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:49)
  checkers.source.SourceChecker.typeProcess(SourceChecker.java:389)
  com.sun.source.util.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:210)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1273)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1232)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:886)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:845)
  com.sun.tools.javac.main.Main.compile(Main.java:421)
  com.sun.tools.javac.main.Main.compile(Main.java:338)
  com.sun.tools.javac.main.Main.compile(Main.java:329)
  com.sun.tools.javac.Main.compile(Main.java:76)
  com.sun.tools.javac.Main.main(Main.java:61)
  checkers.util.CheckerMain.main(CheckerMain.java:41)
[
```

Original issue reported on code.google.com by `charlie....@gmail.com` on 9 May 2012 at 5:48

Attachments:
- [local_compile.sh](https://storage.googleapis.com/google-code-attachments/checker-framework/issue-145/comment-0/local_compile.sh)
- [TestUnion.java](https://storage.googleapis.com/google-code-attachments/checker-framework/issue-145/comment-0/TestUnion.java)
